epinar/assignment2/ancestors.py

```diff
@@ -0,0 +1,113 @@
+class TreeNode:
+	def __init__(self, data):
+		self.data = data
+		self.left = None
+		self.right = None
+
+# Initiates a tree with an input value. Keeps the nodes in a dictionary for fast reach.
+# Each data(key) value should be unique.
+class BinaryTree:
+	def __init__(self, data):
+		self.tree = {}
+		node = TreeNode(data)
+		self.head = node
+		self.tree[data] = node
```

Comment:
1. I thought the assignment question itself required unique keys. As you mentioned, it restricts the usefulness. As well, first, it will be hard to manipulate the tree, for example removing the head, or adding an additional node in between some nodes. Second, it makes it impossible(very messy and weird) to define the structure of a node, being a tree at the same time, such as in your solution. Is that right?
2. I think I understand the overall point here, I will be applying the idea from now on wherever needed.